### DISCUSSION TITLE
https://theme-hope.vuejs.press/zh/guide/feature/meta.html

### DISCUSSION THREAD

**Discussion by giscus** [vuepress-theme-hope/giscus-discussions#589] [Announcements]
# https://theme-hope.vuejs.press/zh/guide/feature/meta.html

主题通过 [@vuepress/plugin-git][git] 插件在页面底部显示页面的最后更新时间与贡献者，并提供了“编辑此页”按钮的支持。 主题同时根据侧边栏配置提供上一页和下一页的导航按钮。

https://theme-hope.vuejs.press/zh/guide/feature/meta.html

**Comment by leoleor** [comment]
我想知道页面贡献者是怎么获取的，我使用的是默认的，发现既不是github的account也不是name，是我真实姓名？

**Reply by Mister-Hope** [reply]
通过git commit的信息，你在提交中使用了真实姓名

**Comment by Lixuhuilll** [comment]
这个插件似乎不能将页面首次提交时间传递给侧边栏，我的侧边栏排序配置了`sidebarSorter: ["readme", "order", "date-desc", "title", "filename"],`，如果我不手动设置页面的 date，就没法按日期排序。但首页却可以不手动配置日期并排序。

**Reply by Mister-Hope** [reply]
有没有一种可能内置的排序器，只排序用户设定过日期的文章会更合理?

不然在它后面的排序器有什么用呢？

**Reply by Mister-Hope** [reply]
退一步讲，那个排序选项是允许你自己设置函数的，所以说你肯定能自己写一个按照你心意中的方式排序。

**Reply by Lixuhuilll** [reply]
> 退一步讲，那个排序选项是允许你自己设置函数的，所以说你肯定能自己写一个按照你心意中的方式排序。

那我想依赖git插件生成的日期，应该怎么编写函数呢？要怎么获取git插件提供的信息？毕竟我都用git插件来管理日期了，再自己给每篇文章写日期，给我一种脱裤子放屁的感觉。。。。。。

**Comment by Alex6357** [comment]
请问 editLink 显示的文本是不能直接更改吗？只能用 metaLocales 间接更改？

**Reply by Mister-Hope** [reply]
什么叫直接什么叫间接？

**Comment by chua-n** [comment]
文档中说：

> 插件会通过页面文件最后一次 git 提交的 UNIX 时间戳 (ms) 来自动生成页面创建时间和最后更新时间，同时根据提交记录生成贡献者。

页面的**创建时间**不应该取决于**第一次**git提交的UNIX时间戳吗？现在我的站点显示的创建时间和更新时间永远都保持一致了，不太对吧

**Reply by Mister-Hope** [reply]
这里只是文档的描述不准确。

**Reply by Mister-Hope** [reply]
已经修改

**Reply by chua-n** [reply]
- 根据git生成的创建时间还是不对，依旧显示的是更新时间而非真正的创建时间；
- 文章的更新时间也不对，显示的是本仓库的上次commit的时间，而非该文件对应的上次修改的提交时间

**Reply by Mister-Hope** [reply]
有没有可能是你压根就没有提供完整的历史？

**Reply by chua-n** [reply]
我的github仓库：https://github.com/chua-n/chua-n.github.io ，源代码为 main 分支，编译后的部署代码为 gh-pages 分支。可以对比下某个md文件编译后的html文件：
- https://github.com/chua-n/chua-n.github.io/blob/main/book/Python/matplotlib/%E5%90%84%E7%A7%8D%E5%9B%BE%E4%BB%B6.md
- https://github.com/chua-n/chua-n.github.io/blob/gh-pages/Python/matplotlib/%E5%90%84%E7%A7%8D%E5%9B%BE%E4%BB%B6.html

这样应该有完整的 git 历史了吧。在 html 里，`datePublished`不是我真正的创建日期，`updated_time`也不是我的更新时间。这个该如何解释呢？

**Reply by Mister-Hope** [reply]
> 这样应该有完整的 git 历史了吧。

# ？ 

- https://github.com/chua-n/chua-n.github.io/blob/main/.github/workflows/deploy-gh-pages.yml
- https://ecosystem.vuejs.press/plugins/git.html#git-repository
- https://vuejs.press/guide/deployment.html#github-pages
- https://github.com/actions/checkout?tab=readme-ov-file#usage

**Reply by chua-n** [reply]
原来如此，是我的问题，大意了。感谢大佬指路！

**Reply by chua-n** [reply]
上述问题已解决。这里发现一个新的问题，楼主考虑解决一下。

在 `@vuepress/plugin-git` 插件中，提取到的创建时间并不**严谨**，该创建时间不包含文件发生过重命名之前的时间。比如某文件`foo.md`，于2021年创建，中间经历多次修改提交，于2023年重命名为`bar.md`，此时插件提取到的创建时间为2023年，实际应该为2021年比较合适。

这一点也很好修改，将插件源码 [getCreatedTime.ts](https://github.com/vuepress/ecosystem/blob/main/plugins/plugin-git/src/node/utils/getCreatedTime.ts) 中调用的 `git log` 命令，加上 `--follow` 参数即可。

即，原来通过 `git --no-pager log --diff-filter=A --format=%at filepath` 命令来获取时间戳，改为使用 `git --no-pager log --follow --diff-filter=A --format=%at filepath` 命令即可。

尝试提了一个PR：https://github.com/vuepress/ecosystem/pull/98

**Reply by Mister-Hope** [reply]
LGTM

**Reply by zhuonixian** [reply]
repo 为gitlab 及 repoLabel: "Gitlab"时，
    "/": {
      // navbar
      navbar: zhNavbar,

      // sidebar
      sidebar: zhSidebar,

      footer: "默认页脚",

      displayFooter: true,

      // page meta
      metaLocales: {
        editLink: "在 git-biz 上编辑此页",
      },

metaLocales不生效

**Reply by Mister-Hope** [reply]
1. 不建议在别人已经完结的讨论下，说一个无关的东西

2. 需要提供最小复现，请开一个新issue并提供所需内容

**Reply by Eschngn** [reply]
“文章的更新时间也不对，显示的是本仓库的上次commit的时间，而非该文件对应的上次修改的提交时间”
大佬 这个问题怎么解决的？我现在push一次代码到仓库中，所有文章的最近更新时间都会变成最新的

**Comment by zhuonixian** [comment]
当repo 及repoLabel: "Gitlab"时，metaLocales 配置不生效。
```ts
...
  locales: {
    /**
     * Chinese locale config
     */
    "/": {
      // navbar
      navbar: zhNavbar,

      // sidebar
      sidebar: zhSidebar,

      footer: "默认页脚",

      displayFooter: true,

      // page meta
      metaLocales: {
        editLink: "在 git-biz 上编辑此页",
      },
    },
...
```

**Comment by minliuhua** [comment]
支持编辑链接是gitee的吗
